README.rst
===============
python-pgextras
===============

.. image:: https://badge.fury.io/py/pgextras.png
    :target: http://badge.fury.io/py/pgextras

.. image:: https://travis-ci.org/scottwoodall/python-pgextras.png?branch=master
        :target: https://travis-ci.org/scottwoodall/python-pgextras

.. image:: https://pypip.in/d/pgextras/badge.png
        :target: https://pypi.python.org/pypi/pgextras

.. image:: https://coveralls.io/repos/scottwoodall/python-pgextras/badge.png
        :target: https://coveralls.io/r/scottwoodall/python-pgextras


Unofficial Python port of `Heroku's pgextras <https://github.com/heroku/heroku-pg-extras>`_
that provides various statistics for a Postgres instance. The biggest difference
of this module is that your Postgres database can be anywhere, not just
hosted with Heroku.

* BSD license
* Tested against 2.7 and 3.3
* `Documentation <http://python-pgextras.rtfd.org>`_
* `Github <https://github.com/scottwoodall/python-pgextras>`_

Features
########

Note: pgextras does not format the output as seen in all the examples below. Instead it
returns an iterable that contains namedtuples. For example, the "Total Indexes Size"
method returns the following::

    [
        Record(table='pgbench_accounts', index_size='2208 kB'),
        Record(table='pgbench_tellers', index_size='16 kB'),
        Record(table='pgbench_branches', index_size='16 kB'),
        Record(table='pgbench_history', index_size='0 bytes')
    ]

It's up to you on how to format and present the data. The results below are
formatted to make it easier to understand the data available to you.

Dev Install
###########

This is specific to the ::
    update-sql-statements branch
Clone the repository and change the branch to `update-sql-statements`
cd in to the directory and run ::
    python setup.py install
This should install the pgextras exectutable

Now, you can use any postgres database to run pgextras


Cache Hit
*********
Calculates your cache hit rate (effective databases are at 99% and up)::

    | name           |    ratio |
    |----------------+----------|
    | index hit rate | 0.999888 |
    | table hit rate | 0.999696 |

Index Usage
***********
Calculates your index hit rate (effective databases are at 99% and up)::

    | relname          |   percent_of_times_index_used |   rows_in_table |
    |------------------+-------------------------------+-----------------|
    | pgbench_history  |                               |          149985 |
    | pgbench_accounts |                            99 |          100000 |
    | pgbench_tellers  |                            96 |              10 |
    | pgbench_branches |                            93 |               1 |

Calls
*****
Show 10 most frequently called queries::

    | qry                                     | exec_time      | prop_exec_time | ncalls   | sync_io_time   |
    |-----------------------------------------+----------------+----------------+----------+----------------|
    | BEGIN;                                  | 0:00:00.140968 | 0.0%           | 414000   | 0:00:00        |
    | INSERT INTO pgbench_history (tid, bid.. | 0:00:03.788899 | 0.0%           | 414000   | 0:00:00        |
    | SELECT abalance FROM pgbench_accounts.. | 0:00:04.471754 | 0.0%           | 414000   | 0:00:00        |
    | UPDATE pgbench_accounts SET abalance .. | 0:00:21.798029 | 0.2%           | 414000   | 0:00:00        |
    | END;                                    | 0:00:00.126220 | 0.0%           | 414000   | 0:00:00        |
    | UPDATE pgbench_branches SET bbalance .. | 0:30:00.544749 | 15.9%          | 414000   | 0:00:00        |
    | UPDATE pgbench_tellers SET tbalance =.. | 2:38:45.396566 | 83.9%          | 414000   | 0:00:00        |
    | BEGIN                                   | 0:00:00.002141 | 0.0%           | 212      | 0:00:00        |
    | SELECT pid, application_name AS sourc.. | 0:00:00.039576 | 0.0%           | 77       | 0:00:00        |
    | SELECT exists( SELECT ? FROM pg_exten.. | 0:00:00.001085 | 0.0%           | 43       | 0:00:00        |


Blocking
********
Display queries holding locks other queries are waiting to be released::

    |   blocked_pid | blocking_statement                                                    | blocking_duration       |   blocking_pid | blocked_statement                                                    | blocked_duration |
    |---------------+-----------------------------------------------------------------------+-------------------------+----------------+----------------------------------------------------------------------|------------------|
    |          1688 | UPDATE pgbench_tellers SET tbalance = tbalance + -2309 WHERE tid = 9; | 0:00:00.018450          |           1724 | UPDATE pgbench_tellers SET tbalance = tbalance + -816 WHERE tid = 9; | 0:00:00.034656   |

Outliers
********
Show 10 queries that have longest execution time in aggregate::

    | qry                                     | exec_time      | prop_exec_time | ncalls   | sync_io_time   |
    |-----------------------------------------+----------------+----------------+----------+----------------|
    | UPDATE pgbench_tellers SET tbalance =.. | 2:59:30.137916 | 83.9%          | 467897   | 0:00:00        |
    | UPDATE pgbench_branches SET bbalance .. | 0:33:53.945889 | 15.8%          | 467856   | 0:00:00        |
    | UPDATE pgbench_accounts SET abalance .. | 0:00:25.384166 | 0.2%           | 467897   | 0:00:00        |
    | SELECT abalance FROM pgbench_accounts.. | 0:00:05.086917 | 0.0%           | 467897   | 0:00:00        |
    | INSERT INTO pgbench_history (tid, bid.. | 0:00:04.356031 | 0.0%           | 467848   | 0:00:00        |
    | vacuum pgbench_branches                 | 0:00:00.336647 | 0.0%           | 17       | 0:00:00        |
    | select count(*) from pgbench_accounts ; | 0:00:00.294740 | 0.0%           | 1        | 0:00:00        |
    | BEGIN;                                  | 0:00:00.160855 | 0.0%           | 467897   | 0:00:00        |
    | END;                                    | 0:00:00.142983 | 0.0%           | 467848   | 0:00:00        |
    | SELECT relname, CASE idx_scan WHEN ? .. | 0:00:00.110683 | 0.0%           | 6        | 0:00:00        |

Vacuum Stats
************
Show dead rows and whether an automatic vacuum is expected to be triggered::

    | schema   | table            | last_vacuum      | last_autovacuum   | rowcount   | dead_rowcount   | autovacuum_threshold   |   expect_autovacuum |
    |----------+------------------+------------------+-------------------+------------+-----------------+------------------------+---------------------|
    | public   | pgbench_tellers  | 2014-04-24 20:02 | 2014-04-24 20:03  | 10         | 0               | 52                     |                     |
    | public   | pgbench_branches | 2014-04-24 20:02 | 2014-04-24 20:03  | 1          | 0               | 50                     |                     |
    | public   | pgbench_history  | 2014-04-23 20:45 |                   | 15000      | 0               | 3050                  |                     |
    | public   | pgbench_accounts | 2014-04-23 20:45 |                   | 100000     | 17581          | 20050                 |                     |

Bloat
*****
Table and index bloat in your database ordered by most wasteful::

    | type   | schemaname   | object_name                                    | bloat | waste        |
    |--------+--------------+------------------------------------------------+-------+--------------|
    | table  | public       | pgbench_accounts                               | 1.3   | 3768 kB      |
    | table  | public       | pgbench_tellers                                | 19    | 144 kB       |
    | table  | public       | pgbench_branches                               | 8     | 56 kB        |

Long Running Queries
********************
Show all queries running longer than five minutes by descending duration::

    | pid   |    duration     |                                      query                                           |
    |-------+-----------------+--------------------------------------------------------------------------------------|
    | 19578 | 02:29:11.200129 | EXPLAIN SELECT  "students".* FROM "students" WHERE "students"."id" = 1450645 LIMIT 1 |
    | 19465 | 02:26:05.542653 | EXPLAIN SELECT  "students".* FROM "students" WHERE "students"."id" = 1889881 LIMIT 1 |
    | 19632 | 02:24:46.962818 | EXPLAIN SELECT  "students".* FROM "students" WHERE "students"."id" = 1581884 LIMIT 1 |

Sequence Scans
**************
Show the count of sequential scans by table descending by order::

    | name             |   count |
    |------------------+---------|
    | pgbench_branches |   57086 |
    | pgbench_tellers  |   15595 |
    | pgbench_accounts |       2 |
    | pgbench_history  |       0 |

Unused Indexes
**************
Show unused and almost unused indexes, ordered by their size relative to the
number of index scans. Exclude indexes of very small tables (less than 5
pages), where the planner will almost invariably select a sequential scan,
but may not in the future as the table grows::

    | table               |                       index                | index_size | index_scans |
    |---------------------+--------------------------------------------+------------+-------------|
    | public.grade_levels | index_placement_attempts_on_grade_level_id | 97 MB      |           0 |
    | public.observations | observations_attrs_grade_resources         | 33 MB      |           0 |
    | public.messages     | user_resource_id_idx                       | 12 MB      |           0 |

Total Table Size
****************
Show the size of the tables (including indexes), descending by size::

    | name             | size    |
    |------------------+---------|
    | pgbench_accounts | 18 MB   |
    | pgbench_history  | 2904 kB |
    | pgbench_tellers  | 272 kB  |
    | pgbench_branches | 256 kB  |

Total Indexes Size
******************
Show the total size of all the indexes on each table, descending by size::

    | table            | index_size   |
    |------------------+--------------|
    | pgbench_accounts | 2208 kB      |
    | pgbench_tellers  | 16 kB        |
    | pgbench_branches | 16 kB        |
    | pgbench_history  | 0 bytes      |

Table Size
**********
Show the size of the tables (excluding indexes), descending by size::

    | name             | size    |
    |------------------+---------|
    | pgbench_accounts | 16 MB   |
    | pgbench_history  | 2904 kB |
    | pgbench_tellers  | 256 kB  |
    | pgbench_branches | 240 kB  |

Index Size
**********
Show the size of indexes, descending by size::

    | name                  | size    |
    |-----------------------+---------|
    | pgbench_accounts_pkey | 2208 kB |
    | pgbench_tellers_pkey  | 16 kB   |
    | pgbench_branches_pkey | 16 kB   |

Total Index Size
****************
Show the total size of all indexes::

    | size    |
    |---------|
    | 2240 kB |

Locks
*****
Display queries with active locks::

     | procpid | relname | transactionid | granted |     query_snippet     |       age
     |---------+---------+---------------+---------+-----------------------+-----------------
     | 31776   |         |               | t       | <IDLE> in transaction | 00:19:29.837898
     | 31776   |         |          1294 | t       | <IDLE> in transaction | 00:19:29.837898
     | 31912   |         |               | t       | select * from hello;  | 00:19:17.94259
     | 3443    |         |               | t       |                      +| 00:00:00
     |         |         |               |         | select               +|
     |         |         |               |         | pg_stat_activi        |

Table Indexes Size
******************
Show the total size of all the indexes on each table, descending by size::

    | table            | index_size   |
    |------------------+--------------|
    | pgbench_accounts | 2208 kB      |
    | pgbench_tellers  | 16 kB        |
    | pgbench_branches | 16 kB        |
    | pgbench_history  | 0 bytes      |

PS
**
View active queries with execution time::

    |   pid | source   | running_for             |   waiting | query                                                                    |
    |-------+----------+-------------------------+-----------+--------------------------------------------------------------------------|
    | 28023 | pgbench  | 0:00:00.107013          |         0 | UPDATE pgbench_accounts SET abalance = abalance + 423 WHERE aid = 10736; |
    | 28018 | pgbench  | 0:00:00.017257          |         0 | END;                                                                     |
    | 28015 | pgbench  | 0:00:00.001055          |         1 | UPDATE pgbench_branches SET bbalance = bbalance + -4203 WHERE bid = 1;   |

Version
*******
Get the Postgres server version::

    | version                                                                                                                           |
    |-----------------------------------------------------------------------------------------------------------------------------------|
    | PostgreSQL 9.3.3 on x86_64-apple-darwin13.0.0, compiled by Apple LLVM version 5.0 (clang-500.2.79) (based on LLVM 3.3svn), 64-bit |


HISTORY.rst
.. :changelog:

History
-------

0.2.1 (2018-12-01)
++++++++++++++++++
* Fixed bug that was truncating index names to only 63 characters
* Updated various sql statement to match heroku's version (thanks @mantrala)

0.2.0 (2014-05-06)
++++++++++++++++++

* Added a CLI option. Original idea for this came from `Marek Wywiał
  <https://github.com/onjin/pgextrascli>`_.
* Remove commas from integer output so no additional parsing is needed if
  someone wants to store the results.
* Make API more consistent by not raising an exception when the
  "pg_stat_statement" module is not installed. Instead a list with a single
  namedtuple containing the error message is returned.

0.1.1 (2014-05-01)
++++++++++++++++++

* updated documentation
* increased test coverage

0.1.0 (2014-04-25)
++++++++++++++++++

* First release on PyPI.


tests/test_pgextras.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest

from mock import patch
import psycopg2
import psycopg2.extras

from pgextras import PgExtras, sql_constants as sql


class TestPgextras(unittest.TestCase):
    def setUp(self):
        self.dbname = 'python_pgextras_unittest'
        self.dsn = 'dbname={dbname}'.format(dbname=self.dbname)
        self.conn = psycopg2.connect(
            self.dsn,
            cursor_factory=psycopg2.extras.NamedTupleCursor
        )
        self.cursor = self.conn.cursor()

    def drop_pg_stat_statement(self):
        if self.is_pg_stat_statement_installed():
            statement = "DROP EXTENSION pg_stat_statements"
            self.cursor.execute(statement)
            self.conn.commit()

    def create_pg_stat_statement(self):
        if not self.is_pg_stat_statement_installed():
            statement = "CREATE EXTENSION pg_stat_statements"
            self.cursor.execute(statement)
            self.conn.commit()

    def is_pg_stat_statement_installed(self):
        self.cursor.execute(sql.PG_STAT_STATEMENT)
        results = self.cursor.fetchall()

        return results[0].available

    def test_that_pg_stat_statement_is_installed(self):
        self.create_pg_stat_statement()

        with PgExtras(dsn=self.dsn) as pg:
            self.assertTrue(pg.pg_stat_statement())

    def test_that_pg_stat_statement_is_not_installed(self):
        self.drop_pg_stat_statement()

        with PgExtras(dsn=self.dsn) as pg:
            self.assertFalse(pg.pg_stat_statement())

    def test_methods_have_one_result(self):
        method_names = ['version', 'total_index_size']

        with PgExtras(dsn=self.dsn) as pg:
            for method_name in method_names:
                func = getattr(pg, method_name)
                results = func()

                self.assertTrue(len(results), 1)

    def test_methods_have_two_results(self):
        method_names = ['cache_hit']

        with PgExtras(dsn=self.dsn) as pg:
            for method_name in method_names:
                func = getattr(pg, method_name)
                results = func()

                self.assertTrue(len(results), 2)

    def test_methods_have_three_results(self):
        method_names = ['index_size']

        with PgExtras(dsn=self.dsn) as pg:
            for method_name in method_names:
                func = getattr(pg, method_name)
                results = func()

                self.assertEqual(len(results), 3)

    def test_methods_have_four_results(self):
        method_names = [
            'table_indexes_size', 'index_usage', 'seq_scans',
            'total_table_size', 'table_size', 'total_indexes_size'
        ]

        with PgExtras(dsn=self.dsn) as pg:
            for method_name in method_names:
                func = getattr(pg, method_name)
                results = func()

                self.assertTrue(len(results), 4)

    def test_calls(self):
        with PgExtras(dsn=self.dsn) as pg:
            if pg.is_pg_at_least_nine_two():
                self.create_pg_stat_statement()
                results = pg.calls()
                self.assertTrue(len(results), 10)
            else:
                results = pg.calls()
                self.assertIsNotNone(results[0].error)

    def test_blocking(self):
        statement = """
            UPDATE pgbench_branches
            SET bbalance = bbalance + 10
            WHERE bid = 1
        """

        blocking_conn = psycopg2.connect(
            database=self.dbname,
            cursor_factory=psycopg2.extras.NamedTupleCursor
        )

        blocking_cursor = blocking_conn.cursor()
        blocking_cursor.execute(statement)

        async_conn = psycopg2.connect(
            database=self.dbname,
            cursor_factory=psycopg2.extras.NamedTupleCursor,
            async=1
        )

        psycopg2.extras.wait_select(async_conn)
        async_cursor = async_conn.cursor()
        async_cursor.execute(statement)

        with PgExtras(dsn=self.dsn) as pg:
            results = pg.blocking()

        self.assertEqual(len(results), 1)

    def test_ps(self):
        """
        If the test suite is ran back to back within one second of each other
        there is a race condition and this test has the potential to fail.
        There will be more than one result because the previous test suites run
        of pg_sleep(2) is still in the pg_stat_activity table. There's also a
        race condition that pg.ps() does not return within 2 seconds and the
        sleep is already gone from the pg_stat_activity.
        """

        statement = """
            SELECT pg_sleep(2);
        """

        async_conn = psycopg2.connect(
            database=self.dbname,
            cursor_factory=psycopg2.extras.NamedTupleCursor,
            async=1
        )

        psycopg2.extras.wait_select(async_conn)
        async_cursor = async_conn.cursor()
        async_cursor.execute(statement)

        with PgExtras(dsn=self.dsn) as pg:
            results = pg.ps()

        self.assertEqual(len(results), 1)

    @patch.object(PgExtras, 'is_pg_at_least_nine_two')
    def test_that_pid_column_returns_correct_column_name(self, mockery):
        mockery.return_value = False

        with PgExtras(dsn=self.dsn) as pg:
            self.assertEqual(pg.pid_column,  'procpid')
            mockery.return_value = True
            self.assertEqual(pg.pid_column,  'pid')

    @patch.object(PgExtras, 'is_pg_at_least_nine_two')
    def test_that_query_column_returns_correct_column_name(self, mockery):
        mockery.return_value = False

        with PgExtras(dsn=self.dsn) as pg:
            self.assertEqual(pg.query_column,  'current_query')
            mockery.return_value = True
            self.assertEqual(pg.query_column,  'query')

    @patch.object(PgExtras, 'version')
    def test_parsing_postgres_version_number(self, mockery):
        Record = type('Record', (object, ), {
            'version': 'PostgreSQL 9.3.3 on x86_64-apple-darwin13.0.0'
        })
        mockery.return_value = [Record]

        with PgExtras(dsn=self.dsn) as pg:
            self.assertTrue(pg.is_pg_at_least_nine_two())
            Record.version = 'PostgreSQL 9.1.1 on x86_64-apple-darwin13.0.0'
            pg._is_pg_at_least_nine_two = None
            self.assertFalse(pg.is_pg_at_least_nine_two())

    def test_error_property_exists_for_missing_pg_stat_statement(self):
        with PgExtras(dsn=self.dsn) as pg:
            results = pg.get_missing_pg_stat_statement_error()

            self.assertIsNotNone(results.error)

    def tearDown(self):
        self.cursor.close()
        self.conn.close()

if __name__ == '__main__':
    unittest.main()


pgextras/__init__.py
# -*- coding: utf-8 -*-
import re
from collections import namedtuple

import psycopg2
import psycopg2.extras

from . import sql_constants as sql

__author__ = 'Scott Woodall'
__email__ = 'scott.woodall@gmail.com'
__version__ = '0.2.1'


class PgExtras(object):
    def __init__(self, dsn=None):
        self.dsn = dsn
        self._pg_stat_statement = None
        self._cursor = None
        self._conn = None
        self._is_pg_at_least_nine_two = None

    def __enter__(self):
        """
        The context manager convention is preferred so that if there are ever
        any exceptions the database cursor/connection will always be closed.
        """
        return self

    def __exit__(self, type, value, trace):
        self.close_db_connection()

    @property
    def cursor(self):
        if self._cursor is None:
            self._conn = psycopg2.connect(
                self.dsn,
                cursor_factory=psycopg2.extras.NamedTupleCursor
            )

            self._cursor = self._conn.cursor()

        return self._cursor

    @property
    def query_column(self):
        """
        PG9.2 changed column names.

        :returns: str
        """

        if self.is_pg_at_least_nine_two():
            return 'query'
        else:
            return 'current_query'

    @property
    def pid_column(self):
        """
        PG9.2 changed column names.

        :returns: str
        """

        if self.is_pg_at_least_nine_two():
            return 'pid'
        else:
            return 'procpid'

    def pg_stat_statement(self):
        """
        Some queries require the pg_stat_statement module to be installed.
        http://www.postgresql.org/docs/current/static/pgstatstatements.html

        :returns: boolean
        """

        if self._pg_stat_statement is None:
            results = self.execute(sql.PG_STAT_STATEMENT)
            is_available = results[0].available

            if is_available:
                self._pg_stat_statement = True
            else:
                self._pg_stat_statement = False

        return self._pg_stat_statement

    def get_missing_pg_stat_statement_error(self):
        Record = namedtuple('Record', 'error')
        error = """
            pg_stat_statements extension needs to be installed in the
            public schema first. This extension is only available on
            Postgres versions 9.2 or greater. You can install it by
            adding pg_stat_statements to shared_preload_libraries in
            postgresql.conf, restarting postgres and then running the
            following sql statement in your database:
            CREATE EXTENSION pg_stat_statements;
        """

        return Record(error)

    def is_pg_at_least_nine_two(self):
        """
        Some queries have different syntax depending what version of postgres
        we are querying against.

        :returns: boolean
        """

        if self._is_pg_at_least_nine_two is None:
            results = self.version()
            regex = re.compile("PostgreSQL (\d+\.\d+\.\d+) on")
            matches = regex.match(results[0].version)
            version = matches.groups()[0]

            if version > '9.2.0':
                self._is_pg_at_least_nine_two = True
            else:
                self._is_pg_at_least_nine_two = False

        return self._is_pg_at_least_nine_two

    def close_db_connection(self):
        if self._cursor is not None:
            self._cursor.close()

        if self._conn is not None:
            self._conn.close()

    def execute(self, statement):
        """
        Execute the given sql statement.

        :param statement: sql statement to run
        :returns: list
        """

        # Make the sql statement easier to read in case some of the queries we
        # run end up in the output
        sql = statement.replace('\n', '')
        sql = ' '.join(sql.split())
        self.cursor.execute(sql)

        return self.cursor.fetchall()

    def cache_hit(self):
        """
        Calculates your cache hit rate (effective databases are at 99% and up).

        Record(
            name='index hit rate',
            ratio=Decimal('0.99994503346970922117')
        )

        :returns: list of Records
        """

        return self.execute(sql.CACHE_HIT)

    def index_usage(self):
        """
        Calculates your index hit rate (effective databases are at 99% and up).

        Record(
            relname='pgbench_history',
            percent_of_times_index_used=None,
            rows_in_table=249976
        )

        :returns: list of Records
        """

        return self.execute(sql.INDEX_USAGE)

    def calls(self, truncate=False):
        """
        Show 10 most frequently called queries. Requires the pg_stat_statements
        Postgres module to be installed.

        Record(
            query='BEGIN;',
            exec_time=datetime.timedelta(0, 0, 288174),
            prop_exec_time='0.0%',
            ncalls='845590',
            sync_io_time=datetime.timedelta(0)
        )

        :param truncate: trim the Record.query output if greater than 40 chars
        :returns: list of Records
        """

        if self.pg_stat_statement():
            if truncate:
                select = """
                    SELECT CASE
                        WHEN length(query) < 40
                        THEN query
                        ELSE substr(query, 0, 38) || '..'
                    END AS qry,
                """
            else:
                select = 'SELECT query,'

            return self.execute(sql.CALLS.format(select=select))
        else:
            return [self.get_missing_pg_stat_statement_error()]

    def blocking(self):
        """
        Display queries holding locks other queries are waiting to be
        released.

        Record(
            pid=40821,
            source='',
            running_for=datetime.timedelta(0, 0, 2857),
            waiting=False,
            query='SELECT pg_sleep(10);'
        )

        :returns: list of Records
        """

        return self.execute(
            sql.BLOCKING.format(
                query_column=self.query_column,
                pid_column=self.pid_column
            )
        )

    def outliers(self, truncate=False):
        """
        Show 10 queries that have longest execution time in aggregate. Requires
        the pg_stat_statments Postgres module to be installed.

        Record(
            qry='UPDATE pgbench_tellers SET tbalance = tbalance + ?;',
            exec_time=datetime.timedelta(0, 19944, 993099),
            prop_exec_time='67.1%',
            ncalls='845589',
            sync_io_time=datetime.timedelta(0)
        )

        :param truncate: trim the Record.qry output if greater than 40 chars
        :returns: list of Records
        """

        if self.pg_stat_statement():
            if truncate:
                query = """
                    CASE WHEN length(query) < 40
                        THEN query
                        ELSE substr(query, 0, 38) || '..'
                    END
                """
            else:
                query = 'query'

            return self.execute(sql.OUTLIERS.format(query=query))
        else:
            return [self.get_missing_pg_stat_statement_error()]

    def vacuum_stats(self):
        """
        Show dead rows and whether an automatic vacuum is expected to be
        triggered.

        Record(
            schema='public',
            table='pgbench_tellers',
            last_vacuum='2014-04-29 14:45',
            last_autovacuum='2014-04-29 14:45',
            rowcount='10',
            dead_rowcount='0',
            autovacuum_threshold='52',
            expect_autovacuum=None
        )

        :returns: list of Records
        """

        return self.execute(sql.VACUUM_STATS)

    def bloat(self):
        """
        Table and index bloat in your database ordered by most wasteful.

        Record(
            type='index',
            schemaname='public',
            object_name='pgbench_accounts::pgbench_accounts_pkey',
            bloat=Decimal('0.2'),
            waste='0 bytes'
        )

        :returns: list of Records
        """

        return self.execute(sql.BLOAT)

    def long_running_queries(self):
        """
        Show all queries longer than five minutes by descending duration.

        Record(
            pid=19578,
            duration=datetime.timedelta(0, 19944, 993099),
            query='SELECT * FROM students'
        )

        :returns: list of Records
        """

        if self.is_pg_at_least_nine_two():
            idle = "AND state <> 'idle'"
        else:
            idle = "AND current_query <> '<IDLE>'"

        return self.execute(
            sql.LONG_RUNNING_QUERIES.format(
                pid_column=self.pid_column,
                query_column=self.query_column,
                idle=idle
            )
        )

    def seq_scans(self):
        """
        Show the count of sequential scans by table descending by order.

        Record(
            name='pgbench_branches',
            count=237
        )

        :returns: list of Records
        """

        return self.execute(sql.SEQ_SCANS)

    def unused_indexes(self):
        """
        Show unused and almost unused indexes, ordered by their size relative
        to the number of index scans. Exclude indexes of very small tables
        (less than 5 pages), where the planner will almost invariably select
        a sequential scan, but may not in the future as the table grows.

        Record(
            table='public.grade_levels',
            index='index_placement_attempts_on_grade_level_id',
            index_size='97 MB',
            index_scans=0
        )

        :returns: list of Records
        """

        return self.execute(sql.UNUSED_INDEXES)

    def total_table_size(self):
        """
        Show the size of the tables (including indexes), descending by size.

        Record(
            name='pgbench_accounts',
            size='15 MB'
        )

        :returns: list of Records
        """

        return self.execute(sql.TOTAL_TABLE_SIZE)

    def total_indexes_size(self):
        """
        Show the total size of all the indexes on each table, descending by
        size.

        Record(
            table='pgbench_accounts',
            index_size='2208 kB'
        )

        :returns: list of Records
        """

        return self.execute(sql.TOTAL_INDEXES_SIZE)

    def table_size(self):
        """
        Show the size of the tables (excluding indexes), descending by size.

        :returns: list
        """

        return self.execute(sql.TABLE_SIZE)

    def index_size(self):
        """
        Show the size of indexes, descending by size.

        :returns: list
        """

        return self.execute(sql.INDEX_SIZE)

    def total_index_size(self):
        """
        Show the total size of all indexes.

        Record(
            size='2240 kB'
        )

        :returns: list of Records
        """

        return self.execute(sql.TOTAL_INDEX_SIZE)

    def locks(self):
        """
        Display queries with active locks.

        Record(
            procpid=31776,
            relname=None,
            transactionid=None,
            granted=True,
            query_snippet='select * from hello;',
            age=datetime.timedelta(0, 0, 288174),
        )

        :returns: list of Records
        """

        return self.execute(
            sql.LOCKS.format(
                pid_column=self.pid_column,
                query_column=self.query_column
            )
        )

    def table_indexes_size(self):
        """
        Show the total size of all the indexes on each table, descending by
        size.

        Record(
            table='pgbench_accounts',
            index_size='2208 kB'
        )

        :returns: list of Records
        """

        return self.execute(sql.TABLE_INDEXES_SIZE)

    def ps(self):
        """
        View active queries with execution time.

        Record(
            pid=28023,
            source='pgbench',
            running_for=datetime.timedelta(0, 0, 288174),
            waiting=0,
            query='UPDATE pgbench_accounts SET abalance = abalance + 423;'
        )

        :returns: list of Records
        """

        if self.is_pg_at_least_nine_two():
            idle = "AND state <> 'idle'"
        else:
            idle = "AND current_query <> '<IDLE>'"

        return self.execute(
            sql.PS.format(
                pid_column=self.pid_column,
                query_column=self.query_column,
                idle=idle
            )
        )

    def version(self):
        """
        Get the Postgres server version.

        Record(
            version='PostgreSQL 9.3.3 on x86_64-apple-darwin13.0.0'
        )

        :returns: list of Records
        """

        return self.execute(sql.VERSION)
